// @graphql-hive/gateway@2.1.2/dist/index.cjs
'use strict';

var cli = require('./cli-Bmg2AY5W.cjs');
var logger = require('@graphql-hive/logger');
var gatewayRuntime = require('@graphql-hive/gateway-runtime');
var pubsub = require('@graphql-hive/pubsub');
var nats = require('@graphql-hive/pubsub/nats');
var redis = require('@graphql-hive/pubsub/redis');
var pluginJwtAuth = require('@graphql-mesh/plugin-jwt-auth');
var pluginPrometheus = require('@graphql-mesh/plugin-prometheus');
var pluginRateLimit = require('@graphql-mesh/plugin-rate-limit');
var pluginHttpCache = require('@graphql-mesh/plugin-http-cache');
var pluginSnapshot = require('@graphql-mesh/plugin-snapshot');
var cacheCfwKv = require('@graphql-mesh/cache-cfw-kv');
var cacheRedis = require('@graphql-mesh/cache-redis');
var cacheLocalforage = require('@graphql-mesh/cache-localforage');
var cacheUpstashRedis = require('@graphql-mesh/cache-upstash-redis');
var transportWs = require('@graphql-mesh/transport-ws');
var transportHttpCallback = require('@graphql-mesh/transport-http-callback');
var transportHttp = require('@graphql-mesh/transport-http');
require('node:cluster');
require('node:module');
require('node:os');
require('node:path');
require('@commander-js/extra-typings');
require('@graphql-yoga/render-graphiql');
require('@graphql-mesh/utils');
require('node:fs/promises');
require('node:url');
require('node:fs');
require('node:http');
require('node:https');
require('@graphql-tools/utils');
require('@opentelemetry/sdk-trace-base');
require('@graphql-tools/code-file-loader');
require('@graphql-tools/graphql-file-loader');
require('@graphql-tools/load');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var pluginPrometheus__default = /*#__PURE__*/_interopDefault(pluginPrometheus);
var pluginRateLimit__default = /*#__PURE__*/_interopDefault(pluginRateLimit);
var pluginHttpCache__default = /*#__PURE__*/_interopDefault(pluginHttpCache);
var pluginSnapshot__default = /*#__PURE__*/_interopDefault(pluginSnapshot);
var cacheCfwKv__default = /*#__PURE__*/_interopDefault(cacheCfwKv);
var cacheRedis__default = /*#__PURE__*/_interopDefault(cacheRedis);
var cacheLocalforage__default = /*#__PURE__*/_interopDefault(cacheLocalforage);
var cacheUpstashRedis__default = /*#__PURE__*/_interopDefault(cacheUpstashRedis);
var transportWs__default = /*#__PURE__*/_interopDefault(transportWs);
var transportHttpCallback__default = /*#__PURE__*/_interopDefault(transportHttpCallback);
var transportHttp__default = /*#__PURE__*/_interopDefault(transportHttp);



exports.defaultOptions = cli.defaultOptions;
exports.defineConfig = cli.defineConfig;
exports.enableModuleCachingIfPossible = cli.enableModuleCachingIfPossible;
exports.getBuiltinPluginsFromConfig = cli.getBuiltinPluginsFromConfig;
exports.getCacheInstanceFromConfig = cli.getCacheInstanceFromConfig;
exports.handleNodeWarnings = cli.handleNodeWarnings;
exports.run = cli.run;
Object.defineProperty(exports, "usePrometheus", {
  enumerable: true,
  get: function () { return pluginPrometheus__default.default; }
});
Object.defineProperty(exports, "useRateLimit", {
  enumerable: true,
  get: function () { return pluginRateLimit__default.default; }
});
Object.defineProperty(exports, "useHttpCache", {
  enumerable: true,
  get: function () { return pluginHttpCache__default.default; }
});
Object.defineProperty(exports, "useSnapshot", {
  enumerable: true,
  get: function () { return pluginSnapshot__default.default; }
});
Object.defineProperty(exports, "CloudflareKVCacheStorage", {
  enumerable: true,
  get: function () { return cacheCfwKv__default.default; }
});
Object.defineProperty(exports, "RedisCacheStorage", {
  enumerable: true,
  get: function () { return cacheRedis__default.default; }
});
Object.defineProperty(exports, "LocalForageCacheStorage", {
  enumerable: true,
  get: function () { return cacheLocalforage__default.default; }
});
Object.defineProperty(exports, "UpstashRedisCache", {
  enumerable: true,
  get: function () { return cacheUpstashRedis__default.default; }
});
Object.defineProperty(exports, "WSTransport", {
  enumerable: true,
  get: function () { return transportWs__default.default; }
});
Object.defineProperty(exports, "HTTPCallbackTransport", {
  enumerable: true,
  get: function () { return transportHttpCallback__default.default; }
});
Object.defineProperty(exports, "HTTPTransport", {
  enumerable: true,
  get: function () { return transportHttp__default.default; }
});
Object.keys(logger).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return logger[k]; }
  });
});
Object.keys(gatewayRuntime).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return gatewayRuntime[k]; }
  });
});
Object.keys(pubsub).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return pubsub[k]; }
  });
});
Object.keys(nats).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return nats[k]; }
  });
});
Object.keys(redis).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return redis[k]; }
  });
});
Object.keys(pluginJwtAuth).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return pluginJwtAuth[k]; }
  });
});
Object.keys(pluginPrometheus).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return pluginPrometheus[k]; }
  });
});
